File: src/main/java/me/miki/shindo/api/ws/presence/PresenceTracker.java
package me.miki.shindo.api.ws.presence;

import com.google.gson.JsonObject;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;


public class PresenceTracker {
    private static final Map<String, PresenceUser> online = new ConcurrentHashMap<>();

    public static boolean isOnline(String uuid) {
        return online.containsKey(uuid);
    }

    public void handleMessage(String type, JsonObject payload) {
        if ("user.join".equals(type)) {
            String uuid = payload.get("uuid").getAsString();
            String name = payload.has("name") ? payload.get("name").getAsString() : "Unknown";
            String acct = payload.has("accountType") ? payload.get("accountType").getAsString() : "OFFLINE";
            online.put(uuid, new PresenceUser(uuid, name, acct, System.currentTimeMillis()));
        } else if ("user.leave".equals(type)) {
            String uuid = payload.get("uuid").getAsString();
            online.remove(uuid);
        }
    }

    public Set<String> allOnlineUuids() {
        return online.keySet();
    }

    public PresenceUser get(String uuid) {
        return online.get(uuid);
    }

    public void clear() {
        online.clear();
    }
}

File: src/main/kotlin/com/shindoclient/shindo/api/websocket/presence/PresenceTracker.kt
package com.shindoclient.shindo.api.websocket.presence

import com.google.gson.JsonObject
import java.util.concurrent.ConcurrentHashMap

class PresenceTracker {
    fun handleMessage(
        type: String,
        payload: JsonObject,
    ) {
        when (type) {
            "user.join" -> {
                val uuid = payload.get("uuid").asString
                val name = if (payload.has("name")) payload.get("name").asString else "Unknown"
                val acct = if (payload.has("accountType")) payload.get("accountType").asString else "OFFLINE"
                online[uuid] = PresenceUser(uuid, name, acct, System.currentTimeMillis())
            }

            "user.leave" -> {
                val uuid = payload.get("uuid").asString
                online.remove(uuid)
            }
        }
    }

    fun allOnlineUuids(): Set<String> = online.keys

    fun get(uuid: String): PresenceUser? = online[uuid]

    fun clear() {
        online.clear()
    }

    companion object {
        private val online = ConcurrentHashMap<String, PresenceUser>()

        @JvmStatic
        fun isOnline(uuid: String): Boolean = online.containsKey(uuid)
    }
}
